File: backend/app/models/project.py
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, func, ARRAY
from sqlalchemy.dialects.postgresql import JSONB
from app.db.session import Base


class Project(Base):
    __tablename__ = "projects"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(128), nullable=False)
    description = Column(Text, nullable=True)
    status = Column(
        String(32), nullable=False, default="WIP"
    )  # WIP, Completed, Archived
    source_type = Column(
        String(32), nullable=False, default="manual"
    )  # manual, github, gitlab
    source_url = Column(String(255), nullable=True)  # GitHub/GitLab URL
    github_username = Column(String(255), nullable=True)  # For GitHub integration
    github_repo = Column(String(255), nullable=True)  # For GitHub integration
    live_url = Column(String(255), nullable=True)  # URL to live demo
    thumbnail_url = Column(String(255), nullable=True)  # Project thumbnail/image
    details = Column(JSONB, nullable=True)  # Additional project details
    order = Column(Integer, nullable=False, default=0)  # Order in projects list
    is_visible = Column(Boolean, default=True)  # To show/hide projects
    
    # GitHub specific fields
    stars_count = Column(Integer, default=0)  # Number of stars
    forks_count = Column(Integer, default=0)  # Number of forks
    watchers_count = Column(Integer, default=0)  # Number of watchers
    language = Column(String(50), nullable=True)  # Primary programming language
    topics = Column(ARRAY(String), nullable=True)  # Repository topics/tags
    last_updated = Column(DateTime(timezone=True), nullable=True)  # Last update time
    homepage_url = Column(String(255), nullable=True)  # Project homepage
    open_issues_count = Column(Integer, default=0)  # Number of open issues
    default_branch = Column(String(50), nullable=True)  # Default branch name
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(
        DateTime(timezone=True), onupdate=func.now(), server_default=func.now()
    )

File: frontend/src/shared/types/project.ts
export interface Project {
  id: number;
  name: string;
  description: string;
  thumbnail_url: string;
  source_url: string;
  live_url?: string;
  topics: string[];
  status: 'WIP' | 'Completed' | 'Archived';
  source_type: 'manual' | 'github' | 'gitlab';
  github_username?: string;
  github_repo?: string;
  stars_count?: number;
  forks_count?: number;
  watchers_count?: number;
  language?: string;
  last_updated?: string;
  homepage_url?: string;
  open_issues_count?: number;
  created_at: string;
  updated_at: string;
}

export type ViewMode = 'grid' | 'list';
export type SortField = 'name' | 'created_at' | 'updated_at' | 'last_updated';
export type SortOrder = 'asc' | 'desc'; 